lib/spack/spack/solver/asp.py

```diff
@@ -3462,102 +3424,46 @@ def value(self) -> str:
+                tty.debug(f"the spec {c} does not provide a C compiler.")
 
-            if using_libc_compatibility() and not c.default_libc:
-                warnings.warn(
-                    f"cannot detect libc from {c.spec}. The compiler will not be used "
-                    f"during concretization."
-                )
-                continue
+            continue
 
-            target = c.target if c.target != "any" else None
-            candidate = KnownCompiler(
-                spec=c.spec, os=c.operating_system, target=target, available=True, compiler_obj=c
+        if using_libc_compatibility() and not CompilerPropertyDetector(c).default_libc():
+            rejected.add(c)
+            warnings.warn(
+                f"cannot detect libc from {c}. The compiler will not be used "
+                f"during concretization."
             )
-            if candidate in self.compilers:
-                warnings.warn(
-                    f"duplicate found for {c.spec} on {c.operating_system}/{c.target}. "
-                    f"Edit your compilers.yaml configuration to remove it."
-                )
-                continue
-
-            self.compilers.add(candidate)
-
-    def with_input_specs(self, input_specs: List[spack.spec.Spec]) -> "CompilerParser":
-        """Accounts for input specs when building the list of possible compilers.
-
-        Args:
-            input_specs: specs to be concretized
-        """
-        strict = spack.concretize.CHECK_COMPILER_EXISTENCE
-        default_os = str(spack.platforms.host().default_os)
-        default_target = str(archspec.cpu.host().family)
-        for s in traverse.traverse_nodes(input_specs):
-            # we don't need to validate compilers for already-built specs
-            if s.concrete or not s.compiler:
-                continue
-
-            version = s.compiler.versions.concrete
-
-            if not version or any(item.spec.satisfies(s.compiler) for item in self.compilers):
-                continue
-
-            # Error when a compiler is not found and strict mode is enabled
-            if strict:
-                raise spack.concretize.UnavailableCompilerVersionError(s.compiler)
+            continue
 
-            # Make up a compiler matching the input spec. This is for bootstrapping.
-            compiler_cls = spack.compilers.class_for_compiler_name(s.compiler.name)
-            compiler_obj = compiler_cls(
-                s.compiler, operating_system=default_os, target=default_target, paths=[None] * 4
-            )
-            self.compilers.add(
-                KnownCompiler(
-                    spec=s.compiler,
-                    os=default_os,
-                    target=default_target,
-                    available=True,
-                    compiler_obj=compiler_obj,
-                )
+        if c in result:
+            warnings.warn(
+                f"duplicate {c.long_spec} compiler found. Edit your packages.yaml to remove it."
```

Comment:
I think we shouldn't warn if the duplicate is coming from layered config.

Reply:
See 05557c8c5ccfea498516c908dd1dc37d282f724f